File: backend/app/schemas/shift_handover.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ── Brief models ─────────────────────────────────────────────────────────────

class ForkliftBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    serial_number: str
    name_en: str


class RentalContractBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    contract_number: str
    status: str


class UserBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    username: str
    full_name: str | None = None


# ── ShiftHandover schemas ────────────────────────────────────────────────────

class ShiftHandoverCreate(BaseModel):
    rental_contract_id: int
    forklift_id: int
    handover_datetime: datetime
    shift_name: str = Field(..., min_length=1, max_length=50)
    handover_person: str = Field(..., min_length=1, max_length=200)
    receiver_person: str = Field(..., min_length=1, max_length=200)
    hour_meter: float = Field(..., ge=0)
    checklist_status: str = Field(default="normal", min_length=1, max_length=30)
    issues_description: str | None = None
    issue_photos: list[str] | None = None
    signatures: dict[str, Any] | None = None


class ShiftHandoverUpdate(BaseModel):
    handover_datetime: datetime | None = None
    shift_name: str | None = Field(default=None, min_length=1, max_length=50)
    handover_person: str | None = Field(default=None, min_length=1, max_length=200)
    receiver_person: str | None = Field(default=None, min_length=1, max_length=200)
    hour_meter: float | None = Field(default=None, ge=0)
    checklist_status: str | None = Field(default=None, min_length=1, max_length=30)
    issues_description: str | None = None
    issue_photos: list[str] | None = None
    signatures: dict[str, Any] | None = None


class ShiftHandoverResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    rental_contract: RentalContractBrief
    forklift: ForkliftBrief
    handover_datetime: datetime
    shift_name: str
    handover_person: str
    receiver_person: str
    hour_meter: float
    checklist_status: str
    issues_description: str | None = None
    issue_photos: list[str] | None = None
    signatures: dict[str, Any] | None = None
    creator: UserBrief | None = None
    created_at: datetime
    updated_at: datetime | None = None


class ShiftHandoverListResponse(BaseModel):
    items: list[ShiftHandoverResponse]
    total: int

File: frontend/src/types/shiftHandover.ts
export type ShiftName = 'Morning' | 'Afternoon' | 'Night'
export type ChecklistStatus = 'Normal' | 'Issues Found'

export interface ForkliftBrief {
  id: number
  serial_number: string
  name_en: string
}

export interface RentalContractBrief {
  id: number
  contract_number: string
  status: string
}

export interface UserBrief {
  id: number
  username: string
  full_name: string | null
}

export interface ShiftHandoverSignatures {
  handover_person?: string | null
  receiver_person?: string | null
}

export interface ShiftHandover {
  id: number
  rental_contract: RentalContractBrief
  forklift: ForkliftBrief
  handover_datetime: string
  shift_name: string
  handover_person: string
  receiver_person: string
  hour_meter: number
  checklist_status: string
  issues_description: string | null
  issue_photos: string[] | null
  signatures: ShiftHandoverSignatures | null
  creator: UserBrief | null
  created_at: string
  updated_at: string | null
}

export interface ShiftHandoverListResponse {
  items: ShiftHandover[]
  total: number
}

export interface ShiftHandoverCreate {
  rental_contract_id: number
  forklift_id: number
  handover_datetime: string
  shift_name: string
  handover_person: string
  receiver_person: string
  hour_meter: number
  checklist_status?: string
  issues_description?: string | null
  issue_photos?: string[] | null
  signatures?: ShiftHandoverSignatures | null
}

export interface ShiftHandoverUpdate {
  handover_datetime?: string
  shift_name?: string
  handover_person?: string
  receiver_person?: string
  hour_meter?: number
  checklist_status?: string
  issues_description?: string | null
  issue_photos?: string[] | null
  signatures?: ShiftHandoverSignatures | null
}

export interface ShiftHandoverListParams {
  rental_contract_id?: number
  forklift_id?: number
  skip?: number
  limit?: number
}
